src/tests/nativeaot/SmokeTests/Exceptions/Exceptions.csproj

```diff
@@ -5,6 +5,8 @@
     <CLRTestPriority>0</CLRTestPriority>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <ServerGarbageCollection>true</ServerGarbageCollection>
```

Comment:
@MichalStrehovsky Does it have anything with the `ServerGarbageCollection` property?

Reply:
I would expect ServerGarbageCollection to work. How's the failure mode in the other test (FrameworkStrings) that you're disabling? They might be related.

This config settings should:
* Make it so that ILC is invoked with `--runtimeopt:gcServer=1` command line option
* This in turn should trigger code to embed gcServer=1 into the object file here: https://github.com/dotnet/runtime/blob/055a9f157dc64e77d1264fec32ac5eb396d81f0b/src/coreclr/tools/aot/ILCompiler/Program.cs#L233-L234 or here: https://github.com/dotnet/runtime/blob/055a9f157dc64e77d1264fec32ac5eb396d81f0b/src/coreclr/tools/aot/ILCompiler/Program.cs#L244-L245
* (You can validate by e.g. adding `--runtimeopt:SomeFunnyString=1` to the command line and checking the output .o file if it contains `SomeFunnyString`, and check again in the result of linking)
* This in turn should be read at runtime here: https://github.com/dotnet/runtime/blob/055a9f157dc64e77d1264fec32ac5eb396d81f0b/src/coreclr/nativeaot/Runtime/RhConfig.cpp#L123-L124 - I'd debug where these are used a check how the program is doing when `gcServer` configuration is looked up from these.

Reply:
For the last bullet, put a breakpoint here:

https://github.com/dotnet/runtime/blob/055a9f157dc64e77d1264fec32ac5eb396d81f0b/src/coreclr/nativeaot/Runtime/RhConfig.cpp#L128

And check what the code is doing when `name` is `gcServer`.

Reply:
> How's the failure mode in the other test (FrameworkStrings) that you're disabling

They should work but they are disabled as Xunit wrapper invokes app bundle generated with different object file. 

The `Environment::TryGetIntegerValue` generates correct variable name `DOTNET_gcServer` but the `PalGetEnvironmentVariable` doesn't find it. I assume that https://github.com/dotnet/runtime/blob/055a9f157dc64e77d1264fec32ac5eb396d81f0b/src/coreclr/nativeaot/Runtime/RhConfig.cpp#L123-L124
generate symbols but I wasn't able to find them using `nm` and inspecting the data section. Where they should be emitted?

Reply:
> PalGetEnvironmentVariable doesn't find it

That one is not expected to find it. The logic first looks at environment variables and if they're not found, looks at embedded settings - this is where we're supposed to find the value embedded by the compiler.

> Where they should be emitted?

This should be emitted into the object file generated by ILC. The responsible code is the one I pointed out in the comment above. The flow should be: compiler gets instructed to put it in the object file by MSBuild (you should see the parameter with `gcServer` string), the .o file should have the value, and we should be able to find it at runtime.

Reply:
After using the debug configuration, the settings are properly set and retrieved in the runtime. The `g_heap_type` has value `GC_HEAP_WKS` since the `FEATURE_SVR_GC` macro is not set and the condition returns false.
https://github.com/dotnet/runtime/blob/1aded3f65f6fd13a60eb43c191f4ea8aa27bf5dc/src/coreclr/nativeaot/Runtime/gcheaputilities.h#L79-L89

Reply:
Ah, thank you for root causing.

We build two flavors of the GC/runtime - one with FEATURE_SVR_GC defined, and the other without:

https://github.com/dotnet/runtime/blob/ef900fa9561ee160f8cbfb4ee312a2e5e018b531/src/coreclr/nativeaot/Runtime/Full/CMakeLists.txt#L29-L31

The build pipeline chooses the right library to link to based on user choice (it is workstation GC by default and server GC when opted in):

https://github.com/dotnet/runtime/blob/ef900fa9561ee160f8cbfb4ee312a2e5e018b531/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets#L32-L33

Looks like this is not happening for iDevices.

Server GC has a higher throughput than workstation GC, but it uses more memory and the binary is a bit bigger.

I don't know if it makes sense to enable this opt-in for iDevices, but it's good to know that it's a no-op for now. If it's an easy fix, we might want to allow opting in.

Reply:
There is an issue with the runtime tests pipeline. Currently, there are separate msbuild projects for AOT compilation and app bundling. The problem is that app bundling doesn't take any properties from the project, which results in linking the WorkstationGC (default garbage collector) instead of the intended ServerGC.

A workaround involves explicitly setting the ServerGC property within the pipeline, but I am not sure if it is appropriate fix. I've made a note of this in the https://github.com/dotnet/runtime/issues/84254.

Reply:
>  Currently, there are separate msbuild projects for AOT compilation and app bundling. The problem is that app bundling doesn't take any properties from the project

Ah. We had such scheme for NativeAOT testing initially as well. We very quickly abandoned it because it doesn't scale for the entire test tree - there are many (definitely at least dozens) of tests that set some property in their project and expect it to take effect and they were not taking effect. We had workarounds like this ugly CDATA section:

https://github.com/dotnet/runtime/pull/64738/files#diff-0eda0b4c7b6597ee1870bf777de26fa39b082dd4e6794b2f17b2a78d4344aa33L14-L22

The approach was abandoned because nobody would want to maintain that for the entire test tree.

Reply:
Yes, we plan to move AOT compilation and app bundling to Helix, which should eventually resolve the issue.